data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroInputCodec.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.dataprepper.plugins.codec.avro;
+
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.FileReader;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.codec.InputCodec;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.log.JacksonLog;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+
+
+/**
+ * An implementation of {@link InputCodec} which parses avro records into fields.
+ */
+@DataPrepperPlugin(name="avro", pluginType = InputCodec.class)
+public class AvroInputCodec implements InputCodec {
+    private static final String MESSAGE_FIELD_NAME = "message";
+    private static final String FILE_NAME = "avro-data";
+    private static final String FILE_SUFFIX = ".avro";
+
+    private static final Logger LOG =  LoggerFactory.getLogger(AvroInputCodec.class);
+
+    @Override
+    public void parse(InputStream inputStream, Consumer<Record<Event>> eventConsumer) throws IOException {
+
+
+        Objects.requireNonNull(inputStream);
+        Objects.requireNonNull(eventConsumer);
+        parseAvroStream(inputStream, eventConsumer);
+
+    }
+
+    private void parseAvroStream(final InputStream inputStream, final Consumer<Record<Event>> eventConsumer)throws IOException {
+
+        File tempFile = File.createTempFile(FILE_NAME, FILE_SUFFIX);
+
+        Files.copy(inputStream, tempFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+        try {
+
+            DatumReader<GenericRecord> datumReaderForSchema = new GenericDatumReader<>();
+            DataFileReader<GenericRecord> dataFileReader = new DataFileReader<>(tempFile, datumReaderForSchema);
+            Schema schema = dataFileReader.getSchema();
+            GenericDatumReader<GenericRecord> genericDatumReader = new GenericDatumReader<>(schema);
+
+            FileReader<GenericRecord> fileReader = DataFileReader.openReader(tempFile, genericDatumReader);
+
+            final Map<String, String> eventData = new HashMap<>();
+
+            for (GenericRecord record : fileReader) {
+
+                for(Schema.Field field : schema.getFields()) {
+
+                    eventData.put(field.name(), record.get(field.name()).toString());
```

Comment:
This code is getting the actual value as a String. However, some of these values may be integers, doubles, longs, etc.

```
record.get(field.name()).toString()
```

Looking at the documentation, the `get(String)` method returns an `Object`. You should be able to pass that `Object` directly into the `eventData`.

https://avro.apache.org/docs/current/api/java/org/apache/avro/generic/GenericRecord.html#get-java.lang.String-

Thus, this line probably works.

```
eventData.put(field.name(), record.get(field.name()));
```

Per [Krishna's comments below](https://github.com/opensearch-project/data-prepper/pull/2414/files#r1153634447), if you verify the actual `Event` you can make sure that this actually works as expected.

---

Also, we need to be sure that Avro sub-records works. I'm not familiar with the Avro API. I'm going to suppose that `get(String)` returns another `GenericRecord` when there is a sub-object. If that is the case, you will need to handle this recursively by going deeper into the nested fields in the object.